packages/dataviews/src/hooks/use-form-validity.ts

```diff
@@ -0,0 +1,427 @@
+			fields.filter( ( field ) => fieldIdsToValidate.has( field.id ) )
+		);
+
+		normalizedFields.forEach( ( field ) => {
+			const value = field.getValue( { item } );
+			if (
+				previousValidatedValuesRef.current.hasOwnProperty( field.id ) &&
+				value === previousValidatedValuesRef.current[ field.id ]
+			) {
+				return;
+			}
+			previousValidatedValuesRef.current[ field.id ] = value;
+
+			// Check isValid.required
+			if (
+				field.isValid.required &&
+				isInvalidForRequired( field.type, value )
+			) {
+				const parentFieldId = fieldParentMap.get( field.id );
+				updateFieldValidity(
+					setFormValidity,
+					field.id,
+					{ required: { type: 'invalid' } },
+					parentFieldId
+				);
+				return;
+			}
+
+			// Check isValid.elements
+			if ( field.isValid.elements && field.elements ) {
+				const validValues = field.elements.map(
+					( element ) => element.value
+				);
+				const parentFieldId = fieldParentMap.get( field.id );
+
+				if ( field.type === 'array' ) {
+					// Arrays (all values must be valid):
+					if ( Array.isArray( value ) ) {
+						const allAreValid = value.every( ( arrayItem ) =>
+							validValues.includes( arrayItem )
+						);
+						if ( allAreValid ) {
+							return;
+						}
+						updateFieldValidity(
+							setFormValidity,
+							field.id,
+							{
+								elements: {
+									type: 'invalid',
+									message:
+										'Value must be one of the elements.',
+								},
+							},
+							parentFieldId
+						);
+						return;
+					}
+
+					updateFieldValidity(
+						setFormValidity,
+						field.id,
+						{
+							elements: {
+								type: 'invalid',
+								message: 'Value must be one of the elements.',
+							},
+						},
+						parentFieldId
+					);
+					return;
+				}
+
+				// Single-value fields:
+				const isValid = validValues.includes( value );
+				if ( isValid ) {
+					return;
+				}
+
+				updateFieldValidity(
+					setFormValidity,
+					field.id,
+					{
+						elements: {
+							type: 'invalid',
+							message: 'Value must be one of the elements.',
+						},
+					},
+					parentFieldId
+				);
+				return;
+			}
+
+			// Check isValid.custom (async)
+			if (
+				typeof field.isValid.custom === 'function' &&
+				field.isValid.custom.constructor.name === 'AsyncFunction'
+			) {
+				const parentFieldId = fieldParentMap.get( field.id );
+				const customAsyncError = field.isValid.custom(
+					deepMerge(
+						item,
+						field.setValue( {
+							item,
+							value,
+						} ) as Partial< Item >
+					),
+					field
+				);
+				if ( customAsyncError === null ) {
+					return;
+				}
+
+				updateFieldValidity(
+					setFormValidity,
+					field.id,
+					{
+						custom: {
+							type: 'validating',
+							message: 'Validating...',
+						},
+					},
+					parentFieldId
+				);
+
+				if ( customAsyncError instanceof Promise ) {
+					customAsyncError
+						.then( ( result ) => {
+							if ( result === null ) {
+								updateFieldValidity(
```

Comment:
Minor: We may have a race condition user makes item in state A, the validation of state A is complex and takes 3 seconds, the user then before the 3 seconds changes item to state B, the validation of state B is very fast and takes 200ms, although we are already on state B the promise for state A may execute after and we show the validation message of state B.

I think we can easily solve that by having some isCancelled variable on the promises result we do nothing if the effect isCancelled, on the effect clean we set isCancelled to true.

Reply:
That clean up logic would also solve the sequence of validation messages shown on the video (it would only show the last):

https://github.com/user-attachments/assets/3ceca81e-8381-4006-9e23-6eaf7ec93dae